streams/src/main/java/org/apache/kafka/streams/kstream/internals/TopologyOptimizerImpl.java

```diff
@@ -0,0 +1,379 @@
+                                   final ProcessDetails processDetails) {
+
+        String topic = processDetails.getSourceTopicArray()[0];
+        Deserializer keyDeserializer = getDeserializer(processDetails.consumedKeySerde());
+        Deserializer valDeserializer = getDeserializer(processDetails.consumedValueSerde());
+
+        internalTopologyBuilder.addGlobalStore(processDetails.getStoreBuilder(),
+                                               descendant.name,
+                                               processDetails.getConsumedTimestampExtractor(),
+                                               keyDeserializer,
+                                               valDeserializer,
+                                               topic,
+                                               processDetails.getConnectProcessorName(),
+                                               processDetails.getkTableSource());
+    }
+
+    private void buildKTableKTableJoin(final InternalTopologyBuilder internalTopologyBuilder,
+                                       final KTableJoinGraphNode ktg,
+                                       final ProcessDetails processDetails) {
+
+        internalTopologyBuilder.addProcessor(ktg.joinThisName, ktg.joinThisProcessor, ktg.thisKtableName);
+        internalTopologyBuilder.addProcessor(ktg.joinOtherName, ktg.joinOtherProcessor, ktg.otherKTableName);
+        internalTopologyBuilder.addProcessor(ktg.joinMerggeName, ktg.joinMergeProcessor, ktg.joinThisName, ktg.joinOtherName);
+        internalTopologyBuilder.connectProcessorAndStateStores(ktg.joinThisName, ktg.joinThisStoreNames);
+        internalTopologyBuilder.connectProcessorAndStateStores(ktg.joinOtherName, ktg.joinOtherStoreNames);
+
+        if (processDetails != null && processDetails.getStoreBuilder() != null) {
+            internalTopologyBuilder.addStateStore(processDetails.getStoreBuilder(), ktg.joinMerggeName);
+        } else if (processDetails != null && processDetails.getStoreSupplier() != null) {
+            internalTopologyBuilder.addStateStore(processDetails.getStoreSupplier(), ktg.joinMerggeName);
+        }
+    }
+
+    private void buildStreamKTableJoin(final InternalTopologyBuilder internalTopologyBuilder,
+                                       final StreamsGraphNode descendant,
+                                       final ProcessDetails processDetails) {
+        internalTopologyBuilder.addProcessor(descendant.name(), processDetails.getProcessorSupplier(), descendant.getPredecessorName());
+        internalTopologyBuilder.connectProcessorAndStateStores(descendant.name(), processDetails.getStoreNames());
+        internalTopologyBuilder.connectProcessors(descendant.name(), processDetails.getConnectProcessorName());
+    }
+
+    private void buildSourceSinkNode(final InternalTopologyBuilder internalTopologyBuilder,
+                                     final SourceSinkNode rgn,
```

Comment:
We should avoid using any abbreviations.

Reply:
Ack